test-data/unit/check-overloading.test

```diff
@@ -348,6 +348,50 @@ def foo(t: Union[List[T], T]) -> T:
         return t[0]
     else:
         return t
+
+class Wrapper(Generic[T]):
+    @overload
+    def foo(self, t: List[T]) -> T: ...
+    @overload
+    def foo(self, t: T) -> T: ...
+    def foo(self, t: Union[List[T], T]) -> T:
+        if isinstance(t, list):
+            return t[0]
+        else:
+            return t
+[builtins fixtures/isinstancelist.pyi]
+
+[case testTypeCheckOverloadImplementationTypeVarDifferingUsage2]
+from typing import overload, Union, List, TypeVar, Generic
+
+T = TypeVar('T')
+
+# Note: this is unsafe when T = object
+@overload
+def foo(t: List[T], s: T) -> int: ...  # E: Overloaded function signatures 1 and 2 overlap with incompatible return types
+@overload
+def foo(t: T, s: T) -> str: ...
+def foo(t, s): pass
+
+# TODO: Why are we getting a different error message here?
+# Shouldn't we be getting the same error message?
+class Wrapper(Generic[T]):
+    @overload
+    def foo(self, t: List[T], s: T) -> int: ...  # E: Overloaded function signatures 1 and 2 are partially overlapping: the two signatures may return incompatible types given certain calls
+    @overload
+    def foo(self, t: T, s: T) -> str: ...
+    def foo(self, t, s): pass
+
+class Dummy(Generic[T]): pass
+
+# Same root issue: why does the additional constraint bound T <: T
+# cause the constraint solver to not infer T = object like it did in the
+# first example?
+@overload
```

Comment:
@ilevkivskyi -- before I continue working on these TODOs, I just want to check in with you and see if this is something you think we should actually fix.

For context, here's a sample program -- we get the same output whether we're using mypy master or this PR:

    from typing import *

    T = TypeVar('T')

    # Example 1: No error reported

    @overload
    def f1(x: int) -> int: ...
    @overload
    def f1(x: T) -> T: ...
    def f1(*args, **kwargs): pass


    # Example 2: Overload is reported as being unsafe

    class Wrapper(Generic[T]):
        @overload
        def f3(self, x: int) -> int: ...
        @overload
        def f3(self, x: T) -> T: ...
        def f3(self, *args, **kwargs): pass


    # Example 3: Overload is reported as being unsafe

    class Dummy(Generic[T]): pass

    @overload
    def f2(d: Dummy[T], x: int) -> int: ...
    @overload
    def f2(d: Dummy[T], x: T) -> T: ...
    def f2(*args, **kwargs): pass

Basically, the issue is that example 1 typechecks (as expected) but example 2 doesn't, which seems like surprising and unintuitive behavior.

I think the root issue is that mypy internally represents example 2 in a way that's basically indistinguishable from example 3. Mypy then tries and fails to unify together the two generics due to the invariance issue you pointed out.

So, before I spend more time on this, I wanted to check in and ask a few questions:

1. Do you think example 3 is actually unsafely overlapping? I'm having some difficultly determining whether or not mypy is actually correct here. 

    (One thing to note is that the behavior of example 3 is technically undefined since the first overload is using just a single typevar. That said, modifying the signature to fix that doesn't change the error message.)

2. Do you think it's worth trying to modify mypy so it doesn't report an error on example 2?

Reply:
Here are my short opinions about two questions:
1. I think the third example is fine (as well as other two).
2. I think it is a bug, i.e. mypy should not report an error there, but it is probably a low-priority one. So I propose to just go ahead with whatever is easier and then we can try fixing this if someone will complain.